core/src/main/scala/kafka/server/BrokerRaftEventManager.scala

```diff
@@ -0,0 +1,165 @@
+package kafka.server
+
+import java.util.concurrent.{LinkedBlockingQueue, TimeUnit}
+
+import kafka.coordinator.group.GroupCoordinator
+import kafka.metrics.KafkaMetricsGroup
+import kafka.server.QuotaFactory.QuotaManagers
+import kafka.utils.ShutdownableThread
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.internals.FatalExitError
+import org.apache.kafka.common.message.UpdateMetadataRequestData
+import org.apache.kafka.common.network.ListenerName
+import org.apache.kafka.common.protocol.{ApiKeys, ApiMessage}
+import org.apache.kafka.common.requests.UpdateMetadataRequest
+import org.apache.kafka.common.utils.Time
+import org.apache.kafka.raft.RaftMessage
+
+object BrokerRaftEventManager {
+  val BrokerRaftEventThreadName = "broker-raft-event-thread"
+  val EventQueueTimeMetricName = "EventQueueTimeMs"
+  val EventQueueSizeMetricName = "EventQueueSize"
+}
+
+trait RaftEventProcessor {
+  def startupBeforeAnyProcessing(): Unit
+  def process(event: RaftMessage): Unit
+}
+
+class QueuedEvent(val event: RaftMessage, val enqueueTimeMs: Long) {
+  override def toString: String = {
+    s"QueuedEvent(event=$event, enqueueTimeMs=$enqueueTimeMs)"
+  }
+}
+
+case object StartupBeforeAnyProcessing extends RaftMessage {
```

Comment:
It seems a little weird to have `StartupBeforeAnyProcessing` and `ShutdownEventThread` extend `RaftMessage`.  They are not Raft messages, after all, right?  Maybe we could have something like a union type of
```
sealed trait Event {}
case object StartupEvent extends Event
case object MessageEvent(raftMessage: RaftMessage) extends Event
case object ShutdownEvent extends Event
```